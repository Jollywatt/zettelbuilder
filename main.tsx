import { walk } from "@std/fs"
import render from "preact-render-to-string"

const combos: { [kind: string]: Set<string> } = {
	"Markdown": new Set(["md"]),
	"plain text": new Set(["txt"]),
	"Typst PDF": new Set(["typ", "pdf"]),
}

function detectNoteKind(extensions: Set<string>): string | null {
	for (const [kind, exts] of Object.entries(combos)) {
		if (extensions.symmetricDifference(exts).size == 0) {
			return kind
		}
	}
	return null
}

interface Note {
	name: string
	kind: string
	files: { [ext: string]: string }
}

export async function findNotes(src: string) {
	const iter = walk(src, {
		includeDirs: false,
		match: [/\.note\.\w+$/],
	})

	const filesByName: { [name: string]: { [ext: string]: string } } = {}

	for await (const file of iter) {
		const m = file.name.match(/(.+)\.note\.(.+)/)
		if (m) {
			const [_, name, ext] = m
			if (!(name in filesByName)) filesByName[name] = {}
			filesByName[name][ext] = file.path
		}
	}

	const notes: { [name: string]: Note } = {}

	for (const name in filesByName) {
		const extensions = Object.keys(filesByName[name])
		const kind = detectNoteKind(new Set(extensions))
		if (kind === null) {
			console.error(
				`Couldn't determine kind of note "${name}" with extensions: ${
					extensions.join(", ")
				}`,
			)
			continue
		}
		notes[name] = {
			name: name,
			kind: kind,
			files: filesByName[name],
		}
	}

	return notes
}

function Main() {
	return (
		<div>
			This is my site
		</div>
	)
}
