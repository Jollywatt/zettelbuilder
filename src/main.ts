export { Note, type NoteFolder, Project } from "./analyse.tsx"
export * as Minimal from "./themes/minimal.tsx"
export * as Zetteldocs from "./themes/zetteldocs.tsx"
