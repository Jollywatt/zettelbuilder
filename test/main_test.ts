import { assertEquals } from "@std/assert"
import { assertSnapshot } from "@std/testing/snapshot"

import { findNotes } from "../main.tsx"

Deno.test("find notes", async function (shot) {
	const notes = await findNotes("test")
	await assertSnapshot(shot, notes)
})
