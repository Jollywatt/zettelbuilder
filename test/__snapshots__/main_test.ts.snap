export const snapshot = {};

snapshot[`find notes 1`] = `
{
  a: {
    files: {
      md: "test/example-notes/a.note.md",
    },
    kind: "Markdown",
    name: "a",
  },
  b: {
    files: {
      txt: "test/example-notes/b.note.txt",
    },
    kind: "plain text",
    name: "b",
  },
  c: {
    files: {
      pdf: "test/example-notes/c.note.pdf",
      typ: "test/example-notes/c.note.typ",
    },
    kind: "Typst PDF",
    name: "c",
  },
}
`;
